File: cs/Problems/Day6.cs
namespace AOC2023.Problems;

public class Day6 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => NumberOfWaysToWin(input);

    private static int NumberOfWaysToWin(ReadOnlySpan<char> input)
    {
        Span<int> times = stackalloc int[10];
        Span<int> distances = stackalloc int[10];

        int lineBreak = input.IndexOf(Environment.NewLine);
        int count = InputParser.ParseNumbers(times, input[5..lineBreak], separator: "  ");
        InputParser.ParseNumbers(distances, input[(lineBreak + 11)..], separator: "  ");

        int waysToWinProduct = 0;

        for (int i = 0; i < count; i++)
        {
            int wayToWin = 0;

            for (int j = 1; j < times[i] - 1; j++)
            {
                if (j * (times[i] - j) > distances[i])
                {
                    wayToWin++;
                }
            }

            waysToWinProduct = Math.Max(waysToWinProduct, 1) * Math.Max(wayToWin, 1);
        }

        return waysToWinProduct;
    }
}

File: go/problems/day6.go
package problems

import (
	"strings"
)

type Day6 struct{}

func (d Day6) Solve(input string) int {
	return numberOfWaysToWin(input)
}

func numberOfWaysToWin(input string) int {
	w2winProd := 0

	lines := strings.Split(input, NewLine)
	times := parseNumbers(lines[0][strings.Index(lines[0], ":")+1:], "  ")
	dists := parseNumbers(lines[1][strings.Index(lines[1], ":")+1:], "  ")

	for i, time := range times {
		w2win := 0

		for j := 1; j < time-1; j++ {
			if j*(time-j) > dists[i] {
				w2win++
			}
		}

		w2winProd = max(w2winProd, 1) * max(w2win, 1)
	}

	return w2winProd
}
